src/audio/pipewire.c

```diff
@@ -0,0 +1,277 @@
+{
+    module_state *state = (module_state *)userdata;
+    // a lot of the code below has been taken from the playing a tone pipewire tutorial, see the pipewire documentation for more details
+
+    struct pw_buffer *b;
+    struct spa_buffer *buf;
+    uint32_t *destination_memory, read_index, available_for_loading, must_load_with_silence;
+    int32_t available_samples, number_of_frames;
+    if ((b = pw_stream_dequeue_buffer(state->stream)) == NULL)
+    {
+        pw_log_warn("out of buffers: %m");
+        return;
+    }
+    buf = b->buffer;
+    // pipewire's buffers are mapped directly from the server to the client
+    // As such, if the memory we're given doesn't exist, we have to abort imediately, because doing otherwise most certainly would lead to undefined behaviour
+    if ((destination_memory = buf->datas[0].data) == NULL)
+        return;
+    // in the process callback, we take the samples from the ring buffer we used in the pipewire_play function and send it to pipewire for playback
+    // this will be done using the following two step algorythm:
+    // we first get the read index of the ringbuffer, which also returns the amount available in there
+    // if we have enough data in the ringbuffer, aka if the read index isn't less than requested, then we load all of it at once in pipewire
+    // however, if the remainder between the filled part of the buffer and the value of requested is greatter than 0, then we fill that part of the buffer with silence, while taking everything else, hoping we will have more samples next time
+    available_samples = spa_ringbuffer_get_read_index(&state->rb, &read_index);
+    number_of_frames = buf->datas[0].maxsize / state->stride; // how many frames of audio are in this chunk, aka how many channel_count touples are available this call to on_process. Technically, requested itself could be used, but this is done instead to avoid some very rare, but important, edge cases
+    if (b->requested > 0)
+    {
+        number_of_frames = SPA_MIN(number_of_frames, b->requested);
+    }
+    // if there's something to be read from the ring buffer at this time, we do so now
+    if (available_samples > 0)
+    {
+        available_for_loading = SPA_MIN(available_samples, number_of_frames);
+    }
+    else
+    {
+        available_for_loading = 0;
+    }
+    // if there's anything else to fill and the ring buffer doesn't contain enough at this point, fill the difference with silence
+    must_load_with_silence = number_of_frames - available_for_loading;
+    if (available_for_loading > 0)
+    {
+        spa_ringbuffer_read_data(&state->rb, state->sample_buffer, SAMPLE_BUFFER_SIZE, read_index, destination_memory, available_for_loading);
+        spa_ringbuffer_read_update(&state->rb, read_index + available_for_loading);
+    }
+    if (must_load_with_silence > 0)
+    {
+        memset(destination_memory + available_for_loading, 0, must_load_with_silence);
+    }
```

Comment:
I would also say to surround this with `#if PW_CHECK_VERSION(0, 3, 49)` and `if (b->requested)` : it's only when we cannot meet the `requested` constraint that we have to introduce silence. If we don't have such a constraint or cannot know it, we'd fall back to completely filling the buffer, but that might be way more than we actually need, and spuriously empty the `sample_buffer`, leading to introduce silence while we had ample time to buffer that.

Reply:
we also want to test `if (b->requested)`. If `requested` is 0, we'll try to fill the whole buffer in, which would be excessive and lead to spurious silence

Reply:
fixed in latest commit